src/Google.Api.Gax/PageStreamExtensions.cs

```diff
@@ -0,0 +1,133 @@
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax
+{
+    public static class PageStreamExtensions
+    {
+        public static IEnumerable<TResource> Flatten<TResponse, TResource>(
+            this IPagedEnumerable<TResponse, TResource> source)
+            where TResponse : IEnumerable<TResource> =>
+            source.SelectMany(page => page);
+
+        public static IAsyncEnumerable<TResource> Flatten<TResponse, TResource>(
+            this IPagedAsyncEnumerable<TResponse, TResource> source)
+            where TResponse : IEnumerable<TResource> =>
+            source.SelectMany(page => page.ToAsyncEnumerable());
+
+        private class FixedPageSizeAsyncEnumerable<TResponse, TResource> : IAsyncEnumerable<FixedPageSize<TResource>>
+            where TResponse : IPageResponse<TResource>
+        {
+            private readonly IPagedAsyncEnumerable<TResponse, TResource> _source;
+            private readonly int _pageSize;
+
+            internal FixedPageSizeAsyncEnumerable(
+                IPagedAsyncEnumerable<TResponse, TResource> source, int pageSize)
+            {
+                _source = source;
+                _pageSize = pageSize;
+            }
+
+            public IAsyncEnumerator<FixedPageSize<TResource>> GetEnumerator() =>
+                new Enumerator(_source.GetEnumerator(), _pageSize);
+
+            private class Enumerator : IAsyncEnumerator<FixedPageSize<TResource>>
+            {
+                private readonly IPagedAsyncEnumerator<TResponse> _enumerator;
+                private readonly int _pageSize;
+
+                internal Enumerator(
+                    IPagedAsyncEnumerator<TResponse> enumerator, int pageSize)
+                {
+                    _enumerator = enumerator;
+                    _pageSize = pageSize;
+                }
+
+                public FixedPageSize<TResource> Current { get; private set; }
+
+                public async Task<bool> MoveNext(CancellationToken cancellationToken)
+                {
+                    var items = new List<TResource>(_pageSize);
+                    while (items.Count < _pageSize)
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+                        int requestCount = _pageSize - items.Count;
+                        var done = !(await _enumerator.MoveNext(requestCount, cancellationToken));
+                        if (done)
+                        {
+                            break;
+                        }
+                        int itemsCount = items.Count;
+                        items.AddRange(_enumerator.Current);
+                        if (items.Count - itemsCount > requestCount)
```

Comment:
Surely this should be `if (items.Count > _pageSize)` otherwise any time we actually need two pages, we're going to break. Before fixing this, let's have a test :)